In [1]:
#Install if necessary
#!pip install kmodes
#!pip install statsmodels


In [1]:
#Import libraries
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Load the dataset
df = pd.read_csv('C:/Users/lewis/Documents/COMP6940/GRP 4 Hotel Cancellation Prediction/Data/Raw/hotel_booking.csv')
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  customer_type  \
0                        0                     0       2  ...      Transient   
1                        0                     0       2  ...      Transient   
2                        0                     1       1  ...      Transient   
3                        0                     1       1  ...      Transient   
4                        0                     2       2  ...      Transient   

    adr required_car_parking_spaces total_of_special_requests  \
0   0.0                           0                         0   
1   0.0                           0                         0   
2  75.0                           0                         0   
3  75.0                           0                         0   
4  98.0                           0                         1   

  reservation_status reservation_status_date            name  \
0          Check-Out              2015-07-01   Ernest Barnes   
1          Check-Out              2015-07-01    Andrea Baker   
2          Check-Out              2015-07-02  Rebecca Parker   
3          Check-Out              2015-07-02    Laura Murray   
4          Check-Out              2015-07-03     Linda Hines   

                         email  phone-number       credit_card  
0  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3            Laura_M@gmail.com  364-656-8427  ************5677  
4           LHines@verizon.com  713-226-5883  ************5498  

[5 rows x 36 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [5]:
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date', 'name', 'email',
       'phone-number', 'credit_card'],
      dtype='object')

#### Original Dataset

| Column Name | Description |
|------------|-------------|
| hotel | Type of hotel (e.g., city hotel or resort hotel) |
| is_canceled | Indicates whether the booking was canceled (1) or not (0) |
| lead_time | Number of days between the booking date and the arrival date |
| arrival_date_year | Year of arrival |
| arrival_date_month | Month of arrival |
| arrival_date_week_number | Week number of the arrival date |
| arrival_date_day_of_month | Day of the month of arrival |
| stays_in_weekend_nights | Number of weekend nights (Saturday/Sunday) in the stay |
| stays_in_week_nights | Number of weekday nights (Monday-Friday) in the stay |
| adults | Number of adults in the booking |
| children | Number of children in the booking |
| babies | Number of babies in the booking |
| meal | Type of meal plan booked (e.g., BB - Bed & Breakfast, HB - Half Board) |
| country | Country of the guest's origin |
| market_segment | Market segment the booking belongs to (e.g., online travel agent, direct, corporate) |
| distribution_channel | Booking distribution channel (e.g., direct, travel agent) |
| is_repeated_guest | Indicates if the guest is a returning customer (1) or new (0) |
| previous_cancellations | Number of previous bookings canceled by the customer |
| previous_bookings_not_canceled | Number of previous bookings not canceled by the customer |
| reserved_room_type | Room type originally reserved by the customer |
| assigned_room_type | Room type actually assigned to the customer |
| booking_changes | Number of changes made to the booking |
| deposit_type | Type of deposit (e.g., no deposit, refundable, non-refundable) |
| agent | ID of the travel agent handling the booking (if applicable) |
| company | ID of the company making the booking (if applicable) |
| days_in_waiting_list | Number of days the booking was on the waiting list |
| customer_type | Type of customer. (e.g., transient, contract, group)|
| adr | Average Daily Rate (revenue per occupied room per day) |
| required_car_parking_spaces | Number of parking spaces required by the customer |
| total_of_special_requests | Number of special requests made by the customer (e.g., extra bed, late check-in) |
| reservation_status | Final status of the reservation (e.g., canceled, checked-out, no-show) |
| reservation_status_date | Date when the reservation status was updated |
| name | Name of the customer |
| email | Email address of the customer |
| phone-number | Contact number of the customer |
| credit_card | Credit card number provided for the booking |

The customer_type column in the dataset categorizes guests based on the nature of their booking. It helps in understanding booking patterns and customer behavior. 

-Transient: Guests who book individual stays and do not belong to any group or contract. These bookings are usually short-term and can come from various sources like direct bookings, online travel agents, or walk-ins.

-Transient-party: Similar to transient customers, but they are traveling as part of a small group. They may have separate bookings but are associated with the same trip or purpose.

-Contract: Guests who book under a corporate or business contract. These bookings are usually made through a company or travel agency and often involve fixed rates for a specific period.

-Group: Bookings that are part of a large group, such as for conferences, weddings, or tour groups. These are typically arranged through travel agencies or event planners.

In [6]:
df.shape

(119390, 36)

In [7]:
# Identify and print columns with missing values
null_columns = df.isnull().sum()
null_columns = null_columns[null_columns > 0]
print(null_columns)

children         4
country        488
agent        16340
company     112593
dtype: int64


The dataset has 119390 rows and the column 'company' has 112593 missing values therefore it is being removed from the dataset due to a large quantity of missing data.

In [8]:
# Drop 'company' column from the dataframe
df.drop(['company'], axis=1, inplace=True)
df.isnull().sum()

hotel                                 0
is_canceled                           0
lead_time                             0
arrival_date_year                     0
arrival_date_month                    0
arrival_date_week_number              0
arrival_date_day_of_month             0
stays_in_weekend_nights               0
stays_in_week_nights                  0
adults                                0
children                              4
babies                                0
meal                                  0
country                             488
market_segment                        0
distribution_channel                  0
is_repeated_guest                     0
previous_cancellations                0
previous_bookings_not_canceled        0
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
agent                             16340
days_in_waiting_list                  0


The agent column is being converted from numerical (agent ID) to categorical (agent or no agent) to simplify analysis by focusing on whether bookings involved an agent, rather than which specific agent was involved.

In [9]:
# Convert the 'agent' column to indicate whether an agent was involved in the booking
df['agent'] = df['agent'].apply(lambda x: 'has agent' if not pd.isna(x) else 'no agent')
# Display the count of missing values across all columns
df.isnull().sum()

hotel                               0
is_canceled                         0
lead_time                           0
arrival_date_year                   0
arrival_date_month                  0
arrival_date_week_number            0
arrival_date_day_of_month           0
stays_in_weekend_nights             0
stays_in_week_nights                0
adults                              0
children                            4
babies                              0
meal                                0
country                           488
market_segment                      0
distribution_channel                0
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
reserved_room_type                  0
assigned_room_type                  0
booking_changes                     0
deposit_type                        0
agent                               0
days_in_waiting_list                0
customer_type                       0
adr         

In [10]:
df.dropna(subset=['children'], inplace=True)
df.isnull().sum()

hotel                               0
is_canceled                         0
lead_time                           0
arrival_date_year                   0
arrival_date_month                  0
arrival_date_week_number            0
arrival_date_day_of_month           0
stays_in_weekend_nights             0
stays_in_week_nights                0
adults                              0
children                            0
babies                              0
meal                                0
country                           488
market_segment                      0
distribution_channel                0
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
reserved_room_type                  0
assigned_room_type                  0
booking_changes                     0
deposit_type                        0
agent                               0
days_in_waiting_list                0
customer_type                       0
adr         

In [11]:
# Fill missing values in the 'country' column with the mode (most frequent value)
most_frequent_country = df['country'].mode()[0]
df['country'].fillna(most_frequent_country, inplace=True)
df.isnull().sum()

C:\Users\lewis\AppData\Local\Temp\ipykernel_15772\587247961.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['country'].fillna(most_frequent_country, inplace=True)


hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces 

There were rows where the Adults, Children and Babies columns were all 0. This is not a logical situation since bookings cannot be made for zero guests and thus these rows were removed

In [12]:
no_guests = df[(df['adults'] == 0) & (df['children'] == 0) & (df['babies'] == 0)]
print("Rows where adults, children, and babies are all 0:", len(no_guests))


Rows where adults, children, and babies are all 0: 180


In [13]:
df = df[~((df['adults'] == 0) & (df['children'] == 0) & (df['babies'] == 0))] #Remove rows where Adults, children and babies are all zero

no_guests = df[(df['adults'] == 0) & (df['children'] == 0) & (df['babies'] == 0)]
print("Rows where adults, children, and babies are all 0:", len(no_guests))

Rows where adults, children, and babies are all 0: 0


In [14]:
# Remove rows with 'Undefined' in distribution_channel and meal columns
df = df[(df['distribution_channel'] != 'Undefined') & (df['meal'] != 'Undefined')]

In [15]:
df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  customer_type     adr  \
0                          0       2  ...      Transient    0.00   
1                          0       2  ...      Transient    0.00   
2                          1       1  ...      Transient   75.00   
3                          1       1  ...      Transient   75.00   
4                          2       2  ...      Transient   98.00   
...                      ...     ...  ...            ...     ...   
119385                     5       2  ...      Transient   96.14   
119386                     5       3  ...      Transient  225.43   
119387                     5       2  ...      Transient  157.71   
119388                     5       2  ...      Transient  104.40   
119389                     7       2  ...      Transient  151.20   

       required_car_parking_spaces total_of_special_requests  \
0                                0                         0   
1                                0                         0   
2                                0                         0   
3                                0                         0   
4                                0                         1   
...                            ...                       ...   
119385                           0                         0   
119386                           0                         2   
119387                           0                         4   
119388                           0                         0   
119389                           0                         2   

       reservation_status reservation_status_date                name  \
0               Check-Out              2015-07-01       Ernest Barnes   
1               Check-Out              2015-07-01        Andrea Baker   

In [16]:
df = df.drop(['name','email','phone-number','credit_card'], axis=1)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 118036 entries, 0 to 119389
Data columns (total 31 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           118036 non-null  object 
 1   is_canceled                     118036 non-null  int64  
 2   lead_time                       118036 non-null  int64  
 3   arrival_date_year               118036 non-null  int64  
 4   arrival_date_month              118036 non-null  object 
 5   arrival_date_week_number        118036 non-null  int64  
 6   arrival_date_day_of_month       118036 non-null  int64  
 7   stays_in_weekend_nights         118036 non-null  int64  
 8   stays_in_week_nights            118036 non-null  int64  
 9   adults                          118036 non-null  int64  
 10  children                        118036 non-null  float64
 11  babies                          118036 non-null  int64  
 12  meal                 

In [18]:
#Change datatype for number of children from float to integer
df['children'] = df['children'].astype(int) 

An arrival date column will be created by merging the arrival year, month and day of month columns and converting the column tyoe to datetime. This is done to calculate the time difference (in days) between cancellation date (reservation status date) and the proposed arrival day

In [19]:
# Convert month names to numbers
month_map = {
    'January': '01', 'February': '02', 'March': '03', 'April': '04',
    'May': '05', 'June': '06', 'July': '07', 'August': '08',
    'September': '09', 'October': '10', 'November': '11', 'December': '12'
}
df['arrival_date_month'] = df['arrival_date_month'].map(month_map)

# Create a new column by combining the date components
df['arrival_date'] = pd.to_datetime(df['arrival_date_year'].astype(str) + '-' + 
                                  df['arrival_date_month'].astype(str) + '-' + 
                                  df['arrival_date_day_of_month'].astype(str))


In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 118036 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   hotel                           118036 non-null  object        
 1   is_canceled                     118036 non-null  int64         
 2   lead_time                       118036 non-null  int64         
 3   arrival_date_year               118036 non-null  int64         
 4   arrival_date_month              118036 non-null  object        
 5   arrival_date_week_number        118036 non-null  int64         
 6   arrival_date_day_of_month       118036 non-null  int64         
 7   stays_in_weekend_nights         118036 non-null  int64         
 8   stays_in_week_nights            118036 non-null  int64         
 9   adults                          118036 non-null  int64         
 10  children                        118036 non-null  int64       

In [21]:
# Convert reservation_status_date to datetime
df['reservation_status_date'] = pd.to_datetime(df['reservation_status_date'])

# Calculate the difference in days
df['days_before_arrival'] = (df['arrival_date'] - df['reservation_status_date']).dt.days

df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                      07                        27   
1                      07                        27   
2                      07                        27   
3                      07                        27   
4                      07                        27   
...                   ...                       ...   
119385                 08                        35   
119386                 08                        35   
119387                 08                        35   
119388                 08                        35   
119389                 08                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...      agent  days_in_waiting_list  \
0                          0       2  ...   no agent                     0   
1                          0       2  ...   no agent                     0   
2                          1       1  ...   no agent                     0   
3                          1       1  ...  has agent                     0   
4                          2       2  ...  has agent                     0   
...                      ...     ...  ...        ...                   ...   
119385                     5       2  ...  has agent                     0   
119386                     5       3  ...  has agent                     0   
119387                     5       2  ...  has agent                     0   
119388                     5       2  ...  has agent                     0   
119389                     7       2  ...  has agent                     0   

       customer_type     adr required_car_parking_spaces  \
0          Transient    0.00                           0   
1          Transient    0.00                           0   
2          Transient   75.00                           0   
3          Transient   75.00                           0   
4          Transient   98.00                           0   
...              ...     ...                         ...   
119385     Transient   96.14                           0   
119386     Transient  225.43                           0   
119387     Transient  157.71                           0   
119388     Transient  104.40                           0   
119389     Transient  151.20                           0   

       total_of_special_requests  reservation_status  reservation_status_date  \
0                              0           Check-Out             

##### Dataset after preprocessing

In [22]:
column_info = pd.DataFrame(columns=['Column', 'Type', 'Details'])

# Populate the DataFrame with column information
for col in df.columns:
    if col in ['hotel', 'arrival_date_month', 'meal', 'country', 'market_segment', 
               'distribution_channel', 'reserved_room_type', 'assigned_room_type', 
               'deposit_type', 'customer_type', 'reservation_status']:
        details = f"Categorical feature: {df[col].unique()}"
    elif col in ['lead_time', 'arrival_date_year', 'arrival_date_week_number', 
                'arrival_date_day_of_month', 'stays_in_weekend_nights', 
                'stays_in_week_nights', 'adults', 'children', 'babies', 
                'is_repeated_guest', 'previous_cancellations', 
                'previous_bookings_not_canceled', 'booking_changes', 
                'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 
                'total_of_special_requests', 'days_before_arrival']:
        details = f"Numerical feature: Mean = {df[col].mean():.2f}, Std = {df[col].std():.2f}"
    elif col == 'is_canceled':
        details = f"Target feature: {df[col].unique()}"
    else:
        details = ""
    
    column_info = pd.concat([column_info, pd.DataFrame({
        'Column': [col],
        'Type': [df[col].dtype],
        'Details': [details]
    })], ignore_index=True)

# Display the DataFrame
print("Columns for df:")
column_info


Columns for df:


Column            Type  \
0                            hotel          object   
1                      is_canceled           int64   
2                        lead_time           int64   
3                arrival_date_year           int64   
4               arrival_date_month          object   
5         arrival_date_week_number           int64   
6        arrival_date_day_of_month           int64   
7          stays_in_weekend_nights           int64   
8             stays_in_week_nights           int64   
9                           adults           int64   
10                        children           int64   
11                          babies           int64   
12                            meal          object   
13                         country          object   
14                  market_segment          object   
15            distribution_channel          object   
16               is_repeated_guest           int64   
17          previous_cancellations           int64   
18  previous_bookings_not_canceled           int64   
19              reserved_room_type          object   
20              assigned_room_type          object   
21                 booking_changes           int64   
22                    deposit_type          object   
23                           agent          object   
24            days_in_waiting_list           int64   
25                   customer_type          object   
26                             adr         float64   
27     required_car_parking_spaces           int64   
28       total_of_special_requests           int64   
29              reservation_status          object   
30         reservation_status_date  datetime64[ns]   
31                    arrival_date  datetime64[ns]   
32             days_before_arrival           int64   

                                              Details  
0   Categorical feature: ['Resort Hotel' 'City Hot...  
1                               Target feature: [0 1]  
2      Numerical feature: Mean = 104.28, Std = 107.14  
3       Numerical feature: Mean = 2016.16, Std = 0.71  
4   Categorical feature: ['07' '08' '09' '10' '11'...  
5        Numerical feature: Mean = 27.20, Std = 13.54  
6         Numerical feature: Mean = 15.79, Std = 8.78  
7          Numerical feature: Mean = 0.92, Std = 0.99  
8          Numerical feature: Mean = 2.49, Std = 1.89  
9          Numerical feature: Mean = 1.86, Std = 0.58  
10         Numerical feature: Mean = 0.10, Std = 0.40  
11         Numerical feature: Mean = 0.01, Std = 0.10  
12         Categorical feature: ['BB' 'FB' 'HB' 'SC']  
13  Categorical feature: ['PRT' 'GBR' 'USA' 'ESP' ...  
14  Categorical feature: ['Direct' 'Corporate' 'On...  
15  Categorical feature: ['Direct' 'Corporate' 'TA...  
16         Numerical feature: Mean = 0.03, Std = 0.17  
17         Numerical feature: Mean = 0.09, Std = 0.85  
18         Numerical feature: Mean = 0.14, Std = 1.51  
19  Categorical feature: ['C' 'A' 'D' 'E' 'G' 'F' ...  
20  Categorical feature: ['C' 'A' 'D' 'E' 'G' 'F' ...  
21         Numerical feature: Mean = 0.22, Std = 0.64  
22  Categorical feature: ['No Deposit' 'Refundable...  
23                                                     
24        Numerical feature: Mean = 2.28, Std = 17.50  
25  Categorical feature: ['Transient' 'Contract' '...  
26      Numerical feature: Mean = 102.07, Std = 50.38  
27         Numerical feature: Mean = 0.06, Std = 0.25  
28         Numerical feature: Mean = 0.58, Std = 0.79  
29  Categorical feature: ['Check-Out' 'Canceled' '...  
30                                                     
31                                                     
32       Numerical feature: Mean = 29.93, Std = 70.35

In [23]:
df.to_csv('preprocessed_hotel_booking.csv', index=False)
print("Preprocessed CSV file 'preprocessed_hotel_booking.csv' has been created.")


Preprocessed CSV file 'preprocessed_hotel_booking.csv' has been created.


In [24]:
# Get datetime columns
datetime_cols = df.select_dtypes(include=['datetime', 'datetimetz']).columns
print("Datetime columns:", datetime_cols)

Datetime columns: Index(['reservation_status_date', 'arrival_date'], dtype='object')
